Sources/Plasma/FeatureLib/pfMetalPipeline/ShaderSrc/ShaderTypes.h

```diff
@@ -0,0 +1,223 @@
+};
+
+enum plMetalVertexAttribute
+{
+    /// position of a vertex
+    VertexAttributePosition                             = 0,
+    /// UV of a vertex. Reserves IDs 1-8.
+    VertexAttributeTexcoord                             = 1,
+    /// Normal attribute of a vertex
+    VertexAttributeNormal                               = 9,
+    /// Color attribute of a vertex
+    VertexAttributeColor                                = 10,
+    /// Animation weight of a vertex
+    VertexAttributeWeights                              = 11,
+};
+
+/// Arguments to the shader compiler to control output
+enum plMetalFunctionConstant
+{
+    /// Numbrer of UVs in the FVF vertex layout.
+    FunctionConstantNumUVs                              = 0,
+    /// Number of layers the shader will need to render
+    FunctionConstantNumLayers                           = 1,
+    /// Source type of the material texture. Metal needs to know if the texture will
+    /// be cubic or 2D in advance.  Eight values reserved.
+    FunctionConstantSources                             = 2,
+    /// Blend modes for each of the layers.
+    FunctionConstantBlendModes                          = 10,
+    /// Render flags for each layer. Eight values reserved.
+    FunctionConstantLayerFlags                          = 18,
+    /// Numbrer of weights in the FVF vertex layout.
+    FunctionConstantNumWeights                          = 26,
+};
+
+enum plMetalLayerPassType: uint8_t
+{
+    PassTypeTexture = 1,
+    PassTypeCubicTexture = 2,
+    PassTypeColor = 3
+};
+
+struct plMetalFragmentShaderArgumentBuffer {
+    __fp16 alphaThreshold;
+};
+#ifndef __METAL_VERSION__
+static_assert(std::is_trivial_v<plMetalFragmentShaderArgumentBuffer>, "plMetalFragmentShaderArgumentBuffer must be a trivial type!");
+#endif
+
+struct plMetalShadowCastFragmentShaderArgumentBuffer {
+    bool pointLightCast;
+};
+#ifndef __METAL_VERSION__
+static_assert(std::is_trivial_v<plMetalShadowCastFragmentShaderArgumentBuffer>, "plMetalShadowCastFragmentShaderArgumentBuffer must be a trivial type!");
+#endif
+
+enum plMetalFragmentShaderTextures {
+    FragmentShaderArgumentAttributeTextures = 0,
+    FragmentShaderArgumentAttributeCubicTextures = 8,
+    FragmentShaderArgumentAttributeUniforms = 32
+};
+
+struct plMetalShaderLightSource {
+    simd::float4 position;
+    half4 ambient;
+    half4 diffuse;
+    half4 specular;
+    simd::float3 direction;
+    simd::float4 spotProps; // (falloff, theta, phi)
+    __fp16 constAtten;
+    __fp16 linAtten;
+    __fp16 quadAtten;
+    __fp16 scale;
+};
+#ifndef __METAL_VERSION__
+static_assert(std::is_trivial_v<plMetalShaderLightSource>, "plMetalShaderLightSource must be a trivial type!");
+#endif
+
+typedef struct
+{
+    uint32_t UVWSrc;
+    matrix_float4x4 transform;
+} UVOutDescriptor;
+#ifndef __METAL_VERSION__
+static_assert(std::is_trivial_v<UVOutDescriptor>, "UVOutDescriptor must be a trivial type!");
+#endif
+
+typedef struct
```

Comment:
Can we use C++ structs?

Reply:
Obj-C and Swift both don't have C++ compatibility on by default which is probably why Metal shaders don't commonly use C++ structs in shader headers. But given that Obj-C++ is on I think it should be fine.